### ISSUE TITLE
IOSSound preloading leaks memory on iOS client

### ISSUE THREAD

**Issue by ericnondahl** [libgdx/libgdx#2906]
When looking into a memory leak using XCode Instruments, our total allocations increased significantly each time we loaded an `IOSSound` and then never decreased, even when dispose() was called on the `IOSSound` (I added logging to make sure this was being properly called.)

We use an `AssetManager` to get `IOSSound` objects in our iOS libGDX game, and then play them with `sound.play(volume, 1, 0)`. The `AssetManager` properly handles the Sound lifecycle - they are loaded and then unloaded at the correct times.

Via a process of elimination, I found that calling `preloadEffect` allocates sound memory, and the subsequent `dispose()` does not free it:

    public IOSSound (FileHandle filePath) {
        soundPath = filePath.file().getPath().replace('\\', '/');
		
        // ------ remove this preloading to fix memory leak ------- //
        soundBuffer = OALSimpleAudio.sharedInstance().preloadEffect(soundPath);
		
        channel = OALSimpleAudio.sharedInstance().getChannelSource();
        sourcePool = channel.getSourcePool().getSources();
    }

later in IOSSound:

    @Override
        public void dispose () {
        // ----- should always free sound memory, but doesn’t ------- //
        OALSimpleAudio.sharedInstance().unloadEffect(soundPath);
	}

It may be that when a sound is playing, it does not free memory correctly. Or the `unloadEffect` does not work in general. I'll look into making a PR. I was crunched for time so I was more interested in working around the issue than fixing it in libgdx, but we should definitely fix it. The memory leak here was pretty major, around ~16mb total sound memory leaked per screen that used sound.

**Comment by badlogic** [comment]
Great find, thanks for reporting!

**Comment by Tom-Ski** [comment]
This could be due to the sound still playing, as the can't be removed from cache if the sound is playing. Perhaps just making sure the sound is stopped before disposing will resolve this leak.

**Comment by ericnondahl** [comment]
I've done a bit more testing - while it would be nice if calling `stop()` in the dispose solved the issue, it doesn't look like it. In fact, the leak happens even if the sounds are never played. It seems like `preloadEffect` creates memory which is not returned when using `unloadEffect`. I also tried setting the soundBuffer to null in the dispose - no change there either.

It seems the only fix that is within libGDX that we could do here is removing the preload entirely. That's what we did for our game, and it doesn't seem to have a noticeable impact on sound performance for us. Maybe someone who is more familiar with the ObjectAL library could look into why the memory is not being freed when we call `unloadEffect`.
